cdc/generation.cc

```diff
@@ -834,16 +834,10 @@ future<> generation_service::check_and_repair_cdc_streams() {
                 latest, db_clock::now());
             should_regenerate = true;
         } else {
-            std::unordered_set<dht::token> gen_ends;
-            for (const auto& entry : gen->entries()) {
-                gen_ends.insert(entry.token_range_end);
-            }
-            for (const auto& metadata_token : tmptr->sorted_tokens()) {
-                if (!gen_ends.contains(metadata_token)) {
-                    cdc_log.warn("CDC generation {} missing token {}. Regenerating.", latest, metadata_token);
-                    should_regenerate = true;
-                    break;
-                }
+            if (tmptr->sorted_tokens().size() < gen->entries().size()) {
```

Comment:
Shouldn't we multiply number of tokens times number of shards? It might be easier to check whether any vnode + shard pair has more than one stream id. Alternatively instead of `tmptr->sorted_tokens().size()`, sum up for each token the number of shards its owning node has. You can take the number of shards from snitch or gossip if I remember correctly.

Reply:
> Shouldn't we multiply number of tokens times number of shards?

No, why?
`gen->entries()` should have one entry per token in the ring.

Checking whether shards are covered is harder, IMO we should improve the situation with vnodes/tokens in this PR first, then we can think about shards in a follow-up.

Reply:
Right. Don't mind me. I misunderstood the code.

Reply:
`gen->entries() should have one entry per token in the ring.`

I don't understand. Don't we have a stream per token + shard? Isn't `gen` storing all the streams?

Reply:
`gen->entries()` is a vector of structs, each struct corresponds to a single vnode. Inside the struct you have multiple stream IDs corresponding to shards. So we want `gen->entries.size()` to be the number of vnodes.